### ISSUE TITLE
eps parameter of torch.nn.utils.spectral_norm doesn't work

### ISSUE THREAD

**Issue by ankitkv** [pytorch/pytorch#20192]
## 🐛 Bug

The `eps` parameter of `torch.nn.utils.spectral_norm` doesn't have the intended effect if a weight matrix is zero.

## To Reproduce

```
>>> import torch
>>> from torch.nn.utils import spectral_norm
>>> from torch import nn
>>> a = torch.randn(3,3)
>>> 
>>> m = nn.Linear(3,3)
>>> m(a)
tensor([[-0.3986, -0.1544, -0.2359],
        [-0.8249,  0.7046,  0.4226],
        [-0.0013, -0.0395,  0.3294]], grad_fn=<AddmmBackward>)
>>> m.weight.data.zero_()
>>> m(a)
tensor([[-0.1742,  0.3715,  0.3060],
        [-0.1742,  0.3715,  0.3060],
        [-0.1742,  0.3715,  0.3060]], grad_fn=<AddmmBackward>)
>>> spectral_norm(m)
Linear(in_features=3, out_features=3, bias=True)
>>> m(a)
tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]], grad_fn=<AddmmBackward>)
```

## Expected behavior

The output of `m(a)` should not be `nan`s, since the default `eps` is `1e-12`.

## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.7

**Comment by SsnL** [comment]
copy pasting my reply on slack:

eps is used in computing vector norm for power iteration. but no matter what the u and v eigen vectors are, you always get sigma = u W^T v = 0.
The behavior is following the original chainer impl: https://github.com/pfnet-research/sngan_projection/blob/e84b1a5f604de5fec268f37c3f26478e80b7f475/source/links/sn_linear.py#L65

**Comment by ankitkv** [comment]
In the source code for `spectral_norm`, `eps` is being used in `normalize`, where `max(eps, norm)` is considered as a denominator. I think the purpose of `eps` is exactly to bring numerical stability when norms are very small. I'm not sure following the original chainer impl is a reason for the bug to not be valid.

The `eps` should be used for `sigma` as well, just as it is used within `normalize`. Otherwise there is no point of `eps` if you get `nan`s with or without it for small norm weights.
